recipe_scrapers/vanillaandbean.py

```diff
@@ -25,5 +26,13 @@ def ingredients(self):
     def instructions(self):
         return self.schema.instructions()
 
+    def ingredient_groups(self):
+        return group_ingredients(
+            self.ingredients(),
+            self.soup,
+            ".wprm-recipe-ingredient-group h4",
+            ".wprm-recipe-ingredient",
+        )
```

Comment:
This helper pattern/interface seems to be working fairly well, given that a number of scrapers are using it now.  I haven't seen any cases of sites -- yet -- where it's not applicable, although I guess we might find some of those in future.

Also: in the back of my mind is a vague sense that the helper has a [theoretical runtime](https://en.wikipedia.org/wiki/Big_O_notation) that's order of `O(n^2)` while I think it could probably be refactored into `O(n)` without affecting most sites.. but the size of `n` is low enough for recipes that we probably don't need to worry about this.

Reply:
I am quite pleased with how well it's working :).  There are two examples where it doesn't work: the NIHealthyEating scraper (which kicked all this off in the first place) and there was one of the candidates in #799 (but I can't remember which one right now).

I have been wondering if the helper can be optimised by checking for an exact ingredient match and if there isn't one, falling back to finding the best match. I don't think this would change the `O(n^2)` runtime though.